pkg/plugins/policies/meshtimeout/plugin/v1alpha1/plugin.go

```diff
@@ -25,6 +35,178 @@ func (p plugin) MatchedPolicies(dataplane *core_mesh.DataplaneResource, resource
+	for _, inbound := range dataplane.Spec.Networking.GetInbound() {
+		iface := dataplane.Spec.Networking.ToInboundInterface(inbound)
+
+		listenerKey := xds.InboundListener{
+			Address: iface.DataplaneIP,
+			Port:    iface.DataplanePort,
+		}
+
+		listener, ok := inboundListeners[listenerKey]
+		if !ok {
+			continue
+		}
+
+		rules, ok := fromRules.Rules[listenerKey]
+		if !ok {
+			continue
+		}
+
+		cluster, ok := inboundClusters[envoy_names.GetLocalClusterName(listenerKey.Port)]
+		if !ok {
+			continue
+		}
+
+		protocol := core_mesh.ParseProtocol(inbound.GetProtocol())
+		if err := configure(rules, xds.MeshSubset(), protocol, listener, cluster); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func applyToOutbounds(rules core_xds.ToRules, outboundListeners map[mesh_proto.OutboundInterface]*envoy_listener.Listener, outboundClusters map[string]*envoy_cluster.Cluster, dataplane *core_mesh.DataplaneResource, routing core_xds.Routing) error {
+	for _, outbound := range dataplane.Spec.Networking.GetOutbound() {
+		oface := dataplane.Spec.Networking.ToOutboundInterface(outbound)
+
+		listener, ok := outboundListeners[oface]
+		if !ok {
+			continue
+		}
+
+		serviceName := outbound.GetTagsIncludingLegacy()[mesh_proto.ServiceTag]
+		cluster, ok := outboundClusters[serviceName]
+		if !ok {
+			continue
+		}
+
+		protocol := inferProtocol(routing, serviceName)
+		if err := configure(rules.Rules, xds.MeshService(serviceName), protocol, listener, cluster); err != nil {
+			return err
+		}
+		if serviceName == "test-server" {
+			log.Info("e2e test", "rules", rules.Rules, "protocol", protocol, "listener", listener, "cluster", cluster)
+		}
+	}
+
 	return nil
 }
+
+func applyToGateway(
+	fromRules xds.FromRules, gatewayListeners map[xds.InboundListener]*envoy_listener.Listener, gatewayClusters map[string]*envoy_cluster.Cluster, resources xds_context.ResourceMap, dataplane *core_mesh.DataplaneResource,
+) error {
+	var gateways *core_mesh.MeshGatewayResourceList
+	if rawList := resources[core_mesh.MeshGatewayType]; rawList != nil {
+		gateways = rawList.(*core_mesh.MeshGatewayResourceList)
+	} else {
+		return nil
+	}
+	log.Info("gateways configured")
+	gateway := xds_topology.SelectGateway(gateways.Items, dataplane.Spec.Matches)
+	if gateway == nil {
+		return nil
+	}
+	// TODO remove log
+	log.Info("gateway selected", "gateway", gateway)
+
+	for _, listener := range gateway.Spec.GetConf().GetListeners() {
+		address := dataplane.Spec.GetNetworking().Address
+		port := listener.GetPort()
+		listenerKey := xds.InboundListener{
+			Address: address,
+			Port:    port,
+		}
+		gatewayListener, ok := gatewayListeners[listenerKey]
+		if !ok {
+			continue
+		}
+
+		rules, ok := fromRules.Rules[listenerKey]
+		if !ok {
+			continue
+		}
+
+		// TODO find a way to extract clusters for gateway route
+		cluster, ok := gatewayClusters[""]
```

Comment:
There is a problem with how to extract proper cluster for gateway from `resourcesSet`. It is because [how cluster names are generated](https://github.com/kumahq/kuma/blob/56c651a69abade0deff268a7eee2ddd4418eb5e0/pkg/plugins/runtime/gateway/route/builders.go#L64). Because of how cluster name is build, I have gateway with name `demo-app` listener with name `demo-app:HTTP:80` and cluster with name `demo-app_kuma-demo_svc_5000-41c96d8c484e3ae3 `. 

I talked with @michaelbeaumont about this and he had three suggestions:
1. use `DestinationClusterName` - this is problematic since I am missing most of the data here in plugin to use this method
2. we already annotate the resources with their Origin, perhaps we need a way to annotate them with even more information - At this moment I don't have proposition on what else can be added to resource for easier cluster matching
3. add possibility to plug in directly where those resources are configured - interesting idea, but I am afraid it will be too much work to change it right now.

Does anyone have ideas how to fix this issue and properly extract this cluster?

Reply:
with 1) the idea is to do exactly the same processing of the `MeshGateway` and `MeshGatewayRoute` resources that we do in the `gateway` logic to get the inputs for `DestinationClusterName`. You already have access to the resources in the plugin, you'd "just" need to extract the logic used in the `gateway` generators.

it's probably the only viable option at the moment